### DISCUSSION TITLE
Attempting to use TSP to find Hamiltonian Paths by setting specific weights does not work

### DISCUSSION THREAD

**Discussion by ribponce** [networkx/networkx#6157] [Algorithms]
Hello everyone, I'm trying to find a cheaper method for evaluating any given graph, checking if it has a Hamiltonian path (a path that visits every vertex in a graph exactly once - and does not necessarily go back to starting vertex), and return the first match, if it exists. My current implementation works but it's very slow on larger inputs, so I've been researching different methods to achieve the same result more efficiently. I found something interesting on this website, but so far I was unable to make it work with NetworkX's TSP implementation.
https://mathimages.swarthmore.edu/index.php/Hamiltonian_Path

> However, an algorithm for finding a Hamiltonian path or cycle can also be found from an algorithm for the Traveling Salesman problem. To do this, take a graph G with n vertices and the complete graph with n vertices, called Kn. Since Kn is complete, G is a subgraph of it. So all the edges in G are also contained in the complete graph. Now, if an edge exists in G, we give it a weight of 0 in the complete graph. If it doesn't exist in G, give it a weight 1 in the complete graph. Now, we can apply our assumed Traveling Salesman algorithm to this complete graph. Any Hamiltonian path in the complete graph with a total weight of 0 will be a Hamiltonian path in G. Thus, a Hamiltonian path or cycle can be found with a Traveling Salesman algorithm.

I created a custom Graph from scratch with custom weights as described above. The subgraph weights are set to 0.0 (the graph I'm interested in) and all other edge weights are set to 1.0 (the generated complete graph, - connecting every node to all other nodes). What I'm experiencing is that the algorithm doesn't seem to care about the weights at all, choosing to traverse the edges with weight 1.0 regardless. I don't think I'm doing anything wrong, could it perhaps be a limitation of the current implementation?

Here's is a sample code with my custom graph with specific weights. "greedy" mode should specifically take the weight into consideration, so I'm really out of ideas here.
I have tried the other functions such as traveling_salesman_problem and christofides but couldn't get it to work either.

```

import networkx as nx
import networkx.algorithms.approximation as nx_app

graph = {'0': {'1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 0.0}, '3': {'weight': 1.0}, '4': {'weight': 0.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '1': {'0': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 0.0}, '3': {'weight': 0.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '10': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '11': {'weight': 0.0}, '12': {'weight': 1.0}, '13': {'weight': 0.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 0.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 0.0}, '8': {'weight': 0.0}, '9': {'weight': 1.0}}, '11': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 0.0}, '12': {'weight': 1.0}, '13': {'weight': 0.0}, '14': {'weight': 0.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 0.0}}, '12': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 0.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '13': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 0.0}, '11': {'weight': 0.0}, '12': {'weight': 1.0}, '14': {'weight': 0.0}, '15': {'weight': 0.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '14': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 0.0}, '12': {'weight': 1.0}, '13': {'weight': 0.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '15': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 0.0}, '13': {'weight': 0.0}, '14': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '2': {'0': {'weight': 0.0}, '1': {'weight': 0.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '3': {'weight': 0.0}, '4': {'weight': 0.0}, '5': {'weight': 0.0}, '6': {'weight': 1.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '3': {'0': {'weight': 1.0}, '1': {'weight': 0.0}, '10': {'weight': 0.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 0.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 0.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '4': {'0': {'weight': 0.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 0.0}, '3': {'weight': 1.0}, '5': {'weight': 0.0}, '6': {'weight': 1.0}, '7': {'weight': 0.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '5': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 0.0}, '3': {'weight': 1.0}, '4': {'weight': 0.0}, '6': {'weight': 1.0}, '7': {'weight': 0.0}, '8': {'weight': 1.0}, '9': {'weight': 1.0}}, '6': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 0.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '7': {'weight': 0.0}, '8': {'weight': 1.0}, '9': {'weight': 0.0}}, '7': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 0.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 0.0}, '5': {'weight': 0.0}, '6': {'weight': 0.0}, '8': {'weight': 0.0}, '9': {'weight': 1.0}}, '8': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 0.0}, '11': {'weight': 1.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 1.0}, '7': {'weight': 0.0}, '9': {'weight': 1.0}}, '9': {'0': {'weight': 1.0}, '1': {'weight': 1.0}, '10': {'weight': 1.0}, '11': {'weight': 0.0}, '12': {'weight': 1.0}, '13': {'weight': 1.0}, '14': {'weight': 1.0}, '15': {'weight': 1.0}, '2': {'weight': 1.0}, '3': {'weight': 1.0}, '4': {'weight': 1.0}, '5': {'weight': 1.0}, '6': {'weight': 0.0}, '7': {'weight': 1.0}, '8': {'weight': 1.0}}}

G = nx.from_dict_of_dicts(graph)
path= nx_app.threshold_accepting_tsp(G, "greedy", weight='weight')
print(path)

```

Any help would be immensely appreciated. Thanks!

**Comment by dschult** [comment]
The TSP solvers are heuristic, meta-heuristic, approximate, or whatever other word you want to use. They will not in general find the optimal route exactly. Just a route that is good in much less time than an exhaustive search. It also searches for a cycle. Not a path. So there is no guarantee that it will do well for finding the optimal path.

**Reply by ribponce** [reply]
Thanks for the answer! I understand, but I do not wish or care to find the optimal route exactly. All I care is to know if a route exists, however my question is why does it not seem to care about the defined weights, and if there is a way to force that somehow.

**Reply by dschult** [reply]
The algorithm should care about weights. The threshold algorithm considers small changes in the given cycle, and accepts those changes if the change increases the length by at most the threshold. Thus it is dependent on the initial cycle presented. The greedy solution is pretty good in this case -- something like only 3 edges with weight 1. In fact the threshold algorithm produces exactly the greedy algorithm outcome -- even when the threshold is set to 100. So, it looks like it may be difficult to improve from that starting position.  Maybe it is the best solution? Do you know a better solution for that graph?

If you start with e.g. [str(i) for i in range(16)] + ['0'] instead of "greedy" it produces a different result for sure...

I normally suggest the `nx_app.traveling_salesman_problem()` function -- which adds the extra edges for you and then substitutes for them in the solution. That result involves a fair amount of retracing parts of the graph,so I'm not sure if it is what you want -- and you probably have tried it already.

**Reply by ribponce** [reply]
Yes, there are better solutions. In fact I wouldn't call the output I'm getting a proper solution, like I said it doesn't seem to care about the weights, at least not in a meaningful way. I thought the algorithm would try to minimize the cost.

![image](https://user-images.githubusercontent.com/81909946/199057840-1fb315af-31d1-44d2-9863-c64464896d25.png)
Here is the proposed graph, it is described by the "graph" variable I posted first. 
It outputs the following 
['0', '4', '7', '8', '10', '11', '13', '15', '12', '1', '3', '6', '9', '5', '2', '14', '0']

You will observe it jumps from node 12 to node 1, which is an edge of weight 1 that I'd hope it would avoid. The resulting cost totals 4, meaning it used 4 edges it shouldn't have.
In this case it is indeed a graph with a "dead-end" on node 12, but I assumed if would recognize that and try to place it as either starting or ending point. Nonetheless, I've seen the threshold_accepting_tsp do the same on other simpler graphs, and fail to give good results.

I also experimented with the nx_app.traveling_salesman_problem() function a bit. With the default, generated nx graphs it seem to work, so I'm starting to wonder if my data should be formatted differently? I noticed a nx.graph object prints as something like this (made up values):
{0: {1: {'weight': 1.0}, 2: {'weight': 0.0}}, 1: {0: {'weight': 0.0}, 2: {'weight': 1.0}}, 2: {0: {'weight': 0.0}, 1: {'weight': 1.0}}}
while when I use my data and generate an nx.graph, I get every key inside single quotes:
{'0': {'1': {'weight': 1.0}, '2': {'weight': 0.0}}, '1': {'0': {'weight': 0.0}, '2': {'weight': 1.0}}, '2': {'0': {'weight': 0.0}, '1': {'weight': 1.0}}}
Do you think that could be causing any issues?

Thanks again for taking the time. Best Regards.

**Reply by mjschwenne** [reply]
One reason that `nx_app.traveling_salesman_problem` may not be working is that it fills in the missing edges weight the weight equal to the shortest path distance between the those nodes. That means that the filled edges will have a total weight of zero since all of the edges in the original graph have a weight zero.

**Reply by ribponce** [reply]
That should not be the case because I'm proving it with a complete graph, meaning explicitly defined weights for all pairs of edges possible.

**Reply by dschult** [reply]
Right -- when using the `nx_app.traveling_salesman_problem` code, you should provide only the original graph -- not the one with the edges already added. Then, since that original graph is unweighted, it will provide 1 as the weight for each edge in the graph and the shortest path length between all other pairs of nodes.

The reason `threshold_accepting_tsp` does not end up doing much better than the greedy algorithm (`greedy_tsp`) is that it considers each possible change as a unit and checks whether the change is greater than the threshold. The changes to avoid e.g. the edge 12 to 1, would require multiple edges to implement and in the mean time there wouldn't be a viable solution. So by looking at single path changes, we only search locally in the solution space and don't find the improvements even when they have cost 0.  The algorithm does indeed try to avoid costly edges. But it is restricted to local changes in the path.

For the example you show, what is the better solution?  The solution I get using a greedy method and the `nx_app.traveling_salesman_problem` function is:
`['0', '2', '3', '6', '7', '10', '13', '10', '8', '10', '11', '14', '11', '10', '11', '9', '6', '3', '10', '13', '15', '12', '15', '13', '10', '3', '1', '3', '10', '13', '15', '13', '10', '7', '4','5', '2', '0']`

Also, the integer valued nodes are converted to strings in the dictionary you provide as `graph` in your code. NetworkX doesn't care if nodes are integers or strings so it shouldn't matter. Just be careful not to mix the two or you're likely to get confused. That shouldn't be an issue in this case.

**Reply by dschult** [reply]
Using the complete graph with weighted edges will work with `traveling_salesman_problem`, but the result will almost certainly end up using some weighted edges. If you really want to avoid those edges, just provide the original graph to that function.  It'll handle the adding of edges and replacing them with paths after-the-search is done.

**Reply by ribponce** [reply]
Thanks, I can try to run the algorithm on the sub-graph instead of the complete graph, as you suggested. 
However the solution you get is exactly what I think is wrong with this, see it visits the same nodes multiple times. I'm pretty sure TSP should never do that.
But thanks for clarifying as well regarding the dictionary strings, one thing less to worry about.

**Reply by dschult** [reply]
TSP **has to** visit nodes more than once...  For example on the way to and from node '12' it must pass through nodes '13 and '15 multiple times (TSP looks for a cycle). Any degree 1 node will cause the path to revisit a node. And longer "hairs" will cause revisiting more often.

**Reply by ribponce** [reply]
> TSP has to visit nodes more than once

I'm confused by your statement. The Salesperson Problem by definition is not allowed to visit the same node more than once.
It asks the following question "Given a list of cities and the distances between each pair of cities, what is the shortest possible route that visits each city exactly once and returns to the origin city?".
I understand in this specific graph, since node 12 is a 1-degree vertex as per the sub-graph weights, it will never have an actual solution to the TSP (which is the same as a Hamiltonian Cycle), it has, however, multiple Hamiltonian Paths, and it's what I'm actually after.

**Reply by mjschwenne** [reply]
> The Salesperson Problem by definition is not allowed to visit the same node more than once.

True, however the TSP is also only defined on complete graphs, so when an incomplete graph is passed to the traveling salesman function it will first complete the graph. If the solution uses edges which do not exist in the original graph it will then post process the solution to remove that edge and replace it with the shortest path between those two nodes in the original graph which will result in output that visits some nodes multiple times.

**Reply by ribponce** [reply]
Yes, I understand it. However in this case I'm in fact feeding it a complete graph, where every node is connected to every other node. The only thing is that the edges contained in the "original" graph (which is a sub-graph of the complete graph I'm feeding the function) have weight of 0, and the edges constructed to turn the sub-graph into a complete graph have the weights of 1. I was doing so in hoping that it would return a Hamiltonian path, like I described in the original post.

**Reply by dschult** [reply]
My understanding of the TSP on a network (that is, with an incomplete graph) is that it removes the words "exactly once" in your question.

The TSP methods are only approximate solutions. My understanding is that finding an exact solution using exhaustive search is the method of choice when an exact solution is needed.  Perhaps you are looking for a networkx function named `hamiltonian_path`. We have investigated that feature and haven't come up with a good (non-exhaustive search) algorithm. If you come up with one please let us know. If you end up writing an exhaustive search, we'd be interested to have a PR for adding it.

**Reply by ribponce** [reply]
I tried using networkx's hamiltonian_path function but it requires a directed graph. Perhaps I can try to turn my graphs into directed. I can report back on that. Regarding an exhaustive search algorithm, that's what I have currently (a slightly modified version of something I found online), it uses no external packages, but it starts to take impossibly long on larger graphs, hence my attempt to find alternatives.

**Reply by dschult** [reply]
Warning--the networkx function `hamiltonian_path` only works with tournament graphs (they have a nice structure which allows a fast operation to get the path).

This is a famous problem that expands the time to solve really fast with the size of the graph. So, I'm not surprised that it started taking impossibly long on larger graphs. :(

**Reply by ribponce** [reply]
Thank you. Yes, precisely. I now realize tournament graphs are the same as directed complete graphs - they always have a Hamiltonian path because all nodes are connected to every other node. Even if I turned my graphs H into complete graphs G (H being a sub-graph of G), this method doesn't seem to allow for any sort of weight control per edge nor start or end points, so it would effectively just find paths outside my original sub-graph contained in the complete graph G.
I've been researching this topic a lot in the past weeks and sadly I'm running out of ideas. But there has to be a way :D
